pvlib/solarposition.py

```diff
@@ -1210,3 +1210,84 @@ def hour_angle(times, longitude, equation_of_time):
     )).total_seconds() / 3600. for t in times])
     timezone = times.tz.utcoffset(times).total_seconds() / 3600.
     return 15. * (hours - 12. - timezone) + longitude + equation_of_time / 4.
+
+
+def _hours(times, hour_angle, longitude, equation_of_time):
+    """helper that converts hour angles in degrees into hours"""
+    tz_info = times.tz  # pytz timezone info
+    tz = tz_info.utcoffset(times).total_seconds() / 3600.
+    return (hour_angle - longitude - equation_of_time / 4.) / 15. + 12. + tz
+
+
+def _times(times, hours):
+    """helper converts hours from an array of floats to localized times"""
+    tz_info = times.tz
+    try:
+        hours = hours.values  # make a new reference
+    except AttributeError:
+        pass
+    try:
+        # OLD pandas, "TypeError: Already tz-aware, use tz_convert to convert."
+        times.tz = None
+    except AttributeError:
+        # NEW pandas, "AttributeError: Cannot directly set timezone. Use
+        # tz_localize() or tz_convert() as appropriate"
+        times = times.tz_localize(None)
+    return pd.DatetimeIndex(
+        times.values.astype('datetime64[D]').astype('datetime64[ns]') +
+        (hours * 3600. * 1.e9).astype('timedelta64[ns]'),
+        tz=tz_info)
+
+
+def sunrise_sunset_transit_analytical(times, latitude, longitude, declination,
+                                      equation_of_time):
+    """
+    Analytical calculation of solar sunrise, sunset, and transit.
+
+    .. warning:: The analytic form neglects the effect of atmospheric
+        refraction.
```

Comment:
Probably "geometric" instead of analytic. Also guessing it neglects more than refraction, but haven't looked at the sources.

Reply:
Also spherical earth, sun is represented by a point at its center, ...

Reply:
Here and in the other "analytical"/"geometrical" functions should we give an estimate of the typical magnitude of the errors? Something like "The error depends on location and time of year but is of order 10 minutes."

Reply:
from the [old test](../../../mikofski/pvlib-python/blob/3b0e59757a8f95e766b1d99dfa77509e113f61d2/pvlib/test/test_solarposition.py#L535-L537) the max differences are

* sunrise 0.11 hours
* sunset 0.11 hours
* transit 0.02 hours

this is a rough estimate, only to 2 decimals, I manually iterated the 2nd decimal to see what tolerance I could get.

Also the errors are one sided, ie: for sunrise the geometric estimate is always *earlier* up to 0.11 hours, and down to some positive, non-zero number, so there's always some sunrise error that is a minute or so earlier than the apparent sunrise, and ie: for sunset it's always *later* up to 0.11 hours, and down to a positive, non-zero number, so there's always some sunset error that is a minute or so later than the apparent sunset.